torchqdynamics/utils.py

```diff
@@ -46,10 +46,28 @@ def ket_overlap(x: Tensor, y: Tensor) -> torch.complex:
     return (ket_to_bra(x) @ y).squeeze(-1).sum(-1)
 
 
-def ket_fidelity(x: Tensor, y: Tensor) -> float:
-    r"""Return the fidelity between two state vectors (kets).
+def sqrtm(x: Tensor) -> Tensor:
+    """Compute the square root of a symmetric or Hermitian positive definite matrix.
 
-    The fidelity between two pure states $\varphi$, $\psi$ is defined by their
+    Args:
+        x: Tensor of dimension `(..., n, n)`.
+
+    Returns:
+        Tensor of dimension `(..., n, n)` holding the square root of `x`.
+    """
+    # code copied from
+    # https://github.com/pytorch/pytorch/issues/25481#issuecomment-1032789228
+    L, Q = torch.linalg.eigh(x)
+    zero = torch.zeros((), device=L.device, dtype=L.dtype)
+    threshold = L.max(-1).values * L.size(-1) * torch.finfo(L.dtype).eps
```

Comment:
Instead of `torch.finfo`, I think it's more readable to import `from torch._prims_common import corresponding_real_dtype as to_real` and to run `to_real(L.dtype)`.

FYI, I made a PR on the pytorch github to implement `dtype.to_real` and `dtype.to_complex` methods: https://github.com/pytorch/pytorch/pull/97935 for exactly this feature.